src/picostdlib/tusb.nim

```diff
@@ -219,5 +682,43 @@ template deviceDescriptorStringCallback*(index, langId, body) =
   proc tudDescriptorStringCb(index: byte, langId: uint16): ptr uint16 {.
       exportC: "tud_descriptor_string_cb", codegendecl: "uint16_t const * $2$3".} =
     body
-const
-  TudHidMouseReport*{.importC: "TUD_HID_REPORT_DESC_MOUSETUD_HID_REPORT_DESC_MOUSE(HID_REPORT_ID(REPORT_ID_MOUSE))".}: uint32 = 0
+
+template hidGetReportCb*(instance, reportId, reportType, buffer, reqLen, body) =
+  ## Set the callback to be invoked when an HID GET_REPORT control request is received.
+  ## 
+  ## This is valid for HID device interfaces only. Application must fill buffer
+  ## with a valid input report and return its length. Return zero will cause the
+  ## stack to STALL request.
+  ## 
+  ## This template must be called exactly once in the application code if an
+  ## HID interface is defined (CFG_TUD_HID > 0).
+  proc tudGetReportCb(instance: uint8, reportId: uint8, reportType: HidReportType,
+      buffer: ptr uint8, reqLen: uint16): uint16 {.exportC: "tud_hid_get_report_cb", cdecl.} =
+    body
+
+template hidSetReportCb*(instance, reportId, reportType, buffer, reqLen, body) =
+  ## Set the callback to be invoked when a SET_REPORT control request is received
+  ## or when receiving data on OUT endpoint ( Report ID = 0, Type = 0 )
+  ## 
+  ## This template must be called exactly once in the application code if an
+  ## HID interface is defined (CFG_TUD_HID > 0).
+  proc tudSetReportCb(instance: uint8, reportId: uint8, reportType: HidReportType,
+      buffer: UncheckedArray[byte], reqLen: uint16)
+      # Need to use .codegendecl for the `uint8 const*` typedef
+      {.exportC, codegendecl: "void tud_hid_set_report_cb(uint8_t instance, uint8_t report_id, hid_report_type_t report_type, uint8_t const* buffer, uint16_t bufsize)".} =
+    body
+
+template hidReportCompleteCallback*(itf, report, len, body) =
+  ## Set the callback to be invoked when a HID input report is successfully send to host.
+  ## 
+  ## Application can use this to send the next report
+  ## Note: For composite reports, report[0] is report ID
+  ## This template is optional but must be called at most once.
```

Comment:
Perhaps we should just ensure this is only called once?
```nim
var calledHIDReport {.compileTime.} = 0
template template hidReportCompleteCallback*(itf, report, len, body) =
  static: 
    if calledHIDReport > 0:
      {.error: "called HIDReportCompleteCallback twice".}
    inc calledHIDReport
   ....
```

Reply:
Sure. Both would result in a compile-time error, but at least this way you get a friendly error from nim instead of a cryptic gcc error. Guess we'd need the same for all the callback templates then?